packages/transaction-controller/src/TransactionController.test.ts

```diff
@@ -211,98 +93,27 @@ function buildMockBlockTracker(latestBlockNumber: string): BlockTracker {
   return fakeBlockTracker;
 }
 
-/**
- * Create an object containing mock result callbacks to be used when testing the approval process.
- *
- * @returns The mock result callbacks.
- */
-function buildMockResultCallbacks(): AcceptResultCallbacks {
-  return {
-    success: jest.fn(),
-    error: jest.fn(),
-  };
-}
-
-/**
- * Create a mock controller messenger.
- *
- * @param opts - Options to customize the mock messenger.
- * @param opts.approved - Whether transactions should immediately be approved or rejected.
- * @param opts.delay - Whether to delay approval or rejection until the returned functions are called.
- * @param opts.resultCallbacks - The result callbacks to return when a request is approved.
- * @returns The mock controller messenger.
- */
-function buildMockMessenger({
-  approved,
-  delay,
-  resultCallbacks,
-}: {
-  approved?: boolean;
-  delay?: boolean;
-  resultCallbacks?: AcceptResultCallbacks;
-}): {
-  messenger: TransactionControllerMessenger;
-  approve: () => void;
-  reject: (reason: unknown) => void;
-} {
-  let approve, reject;
-  let promise: Promise<AddResult>;
-
-  if (delay) {
-    promise = new Promise((res, rej) => {
-      // TODO: Replace `any` with type
-      // eslint-disable-next-line @typescript-eslint/no-explicit-any
-      approve = (value?: any) => res({ resultCallbacks, value });
-      reject = rej;
-    });
-  }
-
-  const messenger = {
-    call: jest.fn().mockImplementation(() => {
-      if (approved) {
-        return Promise.resolve({ resultCallbacks });
-      }
-
-      if (delay) {
-        return promise;
-      }
-
-      // eslint-disable-next-line prefer-promise-reject-errors
-      return Promise.reject({
-        code: errorCodes.provider.userRejectedRequest,
-      });
-    }),
-  } as unknown as TransactionControllerMessenger;
-
-  return {
-    messenger,
-    // TODO: Replace `any` with type
-    // eslint-disable-next-line @typescript-eslint/no-explicit-any
-    approve: approve as unknown as (value?: any) => void,
-    reject: reject as unknown as (reason: unknown) => void,
-  };
-}
-
 /**
  * Wait for the controller to emit a transaction finished event.
  *
- * @param controller - The transaction controller to monitor.
+ * @param messenger - The messenger to monitor.
  * @param options - Options to customize the wait.
  * @param options.confirmed - Whether to wait for the transaction to be confirmed or just finished.
  * @returns A promise that resolves with the transaction meta when the transaction is finished.
  */
 function waitForTransactionFinished(
-  controller: TransactionController,
+  messenger: ControllerMessenger<
+    TransactionControllerActions | AllowedActions,
+    TransactionControllerEvents | AllowedEvents
+  >,
   { confirmed = false } = {},
 ): Promise<TransactionMeta> {
   return new Promise((resolve) => {
-    controller.hub.once(
-      `${controller.state.transactions[0].id}:${
```

Comment:
I guess one limitation of using the messaging system for events is we can't easily do single listeners via `once` for a specific transaction since the events are now all static strings.

Meaning this helper is slightly different in that it's waiting for any transaction to be confirmed, rather than the first transaction in the state.

I assume we weren't relying on this if the tests are still passing 😄

Reply:
Good point — we should probably add an equivalent to ControllerMessenger. In the meantime I've fixed this to emulate `once`.